adapters/zmaticoo/zmaticoo.go

```diff
@@ -0,0 +1,151 @@
+package zmaticoo
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/errortypes"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+	"net/http"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+// Builder builds a new instance of the zmaticoo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	return &adapter{
+		endpoint: config.Endpoint,
+	}, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	adapterRequest, errs := a.makeRequest(request)
+	if errs != nil {
+		return nil, errs
+	}
+	return []*adapters.RequestData{adapterRequest}, nil
+
+}
+
+func (a *adapter) makeRequest(request *openrtb2.BidRequest) (*adapters.RequestData, []error) {
+	errs := getZmaticooExt(request)
+	if errs != nil {
+		return nil, errs
+	}
+	err := transform(request)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+	reqBody, err := json.Marshal(request)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqBody,
+		Headers: headers,
+	}, errs
+}
+
+func transform(request *openrtb2.BidRequest) error {
+	for i, imp := range request.Imp {
+		if imp.Native != nil {
+			var nativeRequest map[string]interface{}
+			nativeCopyRequest := make(map[string]interface{})
+			err := json.Unmarshal([]byte(request.Imp[i].Native.Request), &nativeRequest)
+			if err == nil {
+				_, exists := nativeRequest["native"]
+				if exists {
+					continue
+				}
+				nativeCopyRequest["native"] = nativeRequest
+				nativeReqByte, err := json.Marshal(nativeCopyRequest)
+				if err != nil {
+					return err
+				}
+				nativeCopy := *request.Imp[i].Native
+				nativeCopy.Request = string(nativeReqByte)
+				request.Imp[i].Native = &nativeCopy
+			} else {
+				return err
+			}
```

Comment:
It is custom in go to check the err returned from the function first. I would suggest to invert it - that way you won't need the else statement - 
```
if err := json.Unmarshal([]byte(request.Imp[i].Native.Request), &nativeRequest); err != nil {
    return err
}
_, exists := nativeRequest["native"]
if exists {
    continue
}
nativeCopyRequest["native"] = nativeRequest
nativeReqByte, err := json.Marshal(nativeCopyRequest)
if err != nil {
    return err
}
nativeCopy := *request.Imp[i].Native
nativeCopy.Request = string(nativeReqByte)
request.Imp[i].Native = &nativeCopy

Reply:
ok，i have changed it，can you merge to the master

Reply:
@gargcreation1992 please review